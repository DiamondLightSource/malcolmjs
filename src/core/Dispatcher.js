/*
 * React.js Starter Kit
 * Copyright (c) 2014 Konstantin Tarkus (@koistya), KriaSoft LLC.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

import Flux from 'flux';
import PayloadSources from '../constants/PayloadSources';
import assign from 'react/lib/Object.assign';

/**
 * A singleton that operates as the central hub for application updates.
 * For more information visit https://facebook.github.io/flux/
 */
let Dispatcher = assign(new Flux.Dispatcher(), {

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the server.
   */
  handleServerAction(action) {
    var payload = {
      source: PayloadSources.SERVER_ACTION,
      action: action
    };
    this.dispatch(payload);
  },

  /**
   * @param {object} action The details of the action, including the action's
   * type and additional data coming from the view.
   */
  handleViewAction(action) {
    var payload = {
      source: PayloadSources.VIEW_ACTION,
      action: action
    };
    this.dispatch(payload);
  }

});

export default Dispatcher;
