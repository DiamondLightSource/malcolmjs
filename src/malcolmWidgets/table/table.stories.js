import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import ContainedSimpleTable from './table.stories.container';

storiesOf('Widgets/WidgetTable', module).add(
  'Table',
  withInfo(`
  A simple table (containing text).
  `)(() => (
    <ContainedSimpleTable
      labels={['Test', '1', '2', '3']}
      values={{
        Test: ['#1', '#2', '#3'],
        '1': ['a', 'b', 'c'],
        '2': ['d', 'e', 'f'],
        '3': ['g', 'h', 'i'],
      }}
    />
  ))
);
