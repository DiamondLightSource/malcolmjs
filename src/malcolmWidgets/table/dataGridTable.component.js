/* eslint react/no-array-index-key: 0 */
/* eslint no-underscore-dangle: 0 */

import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { emphasize, fade } from '@material-ui/core/styles/colorManipulator';
import IconButton from '@material-ui/core/IconButton';
import Add from '@material-ui/icons/Add';
import { Typography } from '@material-ui/core';
import ReactDataGrid from 'react-data-grid';
import AttributeAlarm, {
  AlarmStates,
} from '../attributeDetails/attributeAlarm/attributeAlarm.component';
import TableWidgetSelector, { getTableWidgetTags } from './widgetSelector';

const styles = theme => ({
  headerLayout: {
    tableLayout: 'fixed',
    width: 'calc(100% - 15px - 32px)',
  },
  footerLayout: {
    width: '100%',
  },
  headerLayoutNoScroll: {
    tableLayout: 'fixed',
    width: 'calc(100% - 32px)',
  },
  tableLayout: {
    tableLayout: 'fixed',
  },
  tableBody: {
    overflowY: 'auto',
    height: 'calc(100% - 75px)',
    width: '100%',
    backgroundColor: theme.palette.background.paper,
  },
  rowFormat: {
    height: '36px',
  },
  incompleteRowFormat: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.5),
    textAlign: 'Center',
    padding: '2px',
  },
  textHeadings: {
    borderLeft: `2px solid ${theme.palette.divider}`,
    textAlign: 'Center',
  },
  selectedRow: {
    backgroundColor: fade(theme.palette.secondary.main, 0.25),
    textAlign: 'Center',
    padding: '2px',
  },
  textBody: {
    backgroundColor: theme.palette.background.default,
    textAlign: 'Center',
    padding: '2px',
  },
  button: {
    width: '24px',
    height: '24px',
    '&:hover': {
      backgroundColor: 'transparent',
    },
  },
});

const isSelectedRow = props =>
  props.selectedRow === props.row
    ? props.classes.selectedRow
    : props.classes.textBody;

const AlarmCell = props => (
  <IconButton
    className={props.classes.button}
    disableRipple
    onClick={() => {
      props.infoClickHandler(props.path, `row.${props.row}`);
    }}
  >
    <AttributeAlarm
      alarmSeverity={
        props.flags.rows[props.row] &&
        (props.flags.rows[props.row]._dirty ||
          props.flags.rows[props.row]._isChanged)
          ? AlarmStates.DIRTY
          : AlarmStates.NO_ALARM
      }
    />
  </IconButton>
);

class NewWidgetTable extends React.Component {
  componentDidMount() {
    const headerCells = [
      ...document.getElementsByClassName('react-grid-HeaderCell'),
    ];
    const header = [...document.getElementsByClassName('react-grid-Header')];
    const canvas = [...document.getElementsByClassName('react-grid-Canvas')];
    if (headerCells.length !== 0) {
      headerCells.forEach(cell => {
        const oldStyle = cell.getAttribute('style');
        cell.setAttribute(
          'style',
          ` ${oldStyle}background-color: ${
            this.props.theme.palette.background.paper
          }; font-family: roboto; border-bottom-width: 0; border-right-width: 2px; border-right-color: ${
            this.props.theme.palette.divider
          } height: 36px; padding: 0; vertical-align: middle; margin-right: 2px`
        );
      });
    }
    if (canvas.length !== 0) {
      canvas.forEach(cell => {
        const oldStyle = cell.getAttribute('style');
        cell.setAttribute(
          'style',
          ` ${oldStyle}background-color: ${
            this.props.theme.palette.background.default
          };`
        );
      });
    }
    if (header.length !== 0) {
      header.forEach(cell => {
        const oldStyle = cell.getAttribute('style');
        cell.setAttribute(
          'style',
          ` ${oldStyle}background-color: ${
            this.props.theme.palette.background.paper
          };`
        );
      });
    }
  }
  render() {
    const columnLabels =
      this.props.localState === undefined
        ? Object.keys(this.props.attribute.raw.meta.elements)
        : this.props.localState.labels;
    const values =
      this.props.localState === undefined
        ? this.props.attribute.raw.value[columnLabels[0]].map((val, row) => {
            const rowData = {};
            columnLabels.forEach(label => {
              rowData[label] = this.props.attribute.raw.value[label][row];
            });
            return rowData;
          })
        : this.props.localState.value;
    const flags =
      this.props.localState === undefined
        ? { rows: [], table: {} }
        : this.props.localState.flags;
    const meta =
      this.props.localState === undefined
        ? this.props.attribute.raw.meta
        : this.props.localState.meta;
    const rowChangeHandler = (rowPath, newValue) => {
      let rowValue;
      if (columnLabels === undefined) {
        rowValue = newValue;
      } else {
        rowValue = {};
        columnLabels.forEach(label => {
          rowValue[label] = values[rowPath.row][label];
          return 0;
        });
        rowValue[rowPath.label] = newValue;
      }
      this.props.eventHandler(
        this.props.attribute.calculated.path,
        rowValue,
        rowPath.row
      );
    };
    const rowFlagHandler = (rowPath, flagType, flagState) => {
      if (this.props.localState !== undefined) {
        const rowFlags =
          this.props.localState.flags.rows[rowPath.row] === undefined
            ? {}
            : this.props.localState.flags.rows[rowPath.row];
        if (rowFlags[flagType] === undefined) {
          rowFlags[flagType] = {};
        }
        rowFlags[flagType][rowPath.label] = flagState;
        rowFlags[`_${flagType}`] = Object.values(rowFlags[flagType]).some(
          val => val
        );
        this.props.setFlag(
          this.props.attribute.calculated.path,
          rowPath.row,
          flagType,
          rowFlags
        );
      }
    };
    const columnWidgetTags = getTableWidgetTags(meta);
    const columnHeadings =
      columnLabels === undefined
        ? [
            {
              key: 'value',
              name: meta.label,
              resizable: true,
            },
          ]
        : columnLabels.map(label => ({
            key: label,
            name: meta.elements[label].label
              ? meta.elements[label].label
              : label,
            resizable: true,
          }));
    if (!this.props.hideInfo) {
      columnHeadings.splice(0, 0, {
        key: 'alarmState',
        name: 'Alarm',
        resizable: true,
      });
    }
    const getRowData = (row, columnWidgetTags, props) => {
      let valueCells;
      if (columnLabels === undefined) {
        valueCells = {
          value: (
            <TableWidgetSelector
              columnWidgetTag={columnWidgetTags[0]}
              value={values[row]}
              rowPath={{ row, column: 0 }}
              rowChangeHandler={rowChangeHandler}
              columnMeta={meta}
              setFlag={rowFlagHandler}
            />
          ),
        };
      } else {
        valueCells = {};
        columnLabels.forEach((label, column) => {
          valueCells[label] = (
            <TableWidgetSelector
              columnWidgetTag={columnWidgetTags[column]}
              value={values[row][label]}
              rowPath={{ label, row, column }}
              rowChangeHandler={rowChangeHandler}
              columnMeta={meta.elements[label]}
              setFlag={rowFlagHandler}
            />
          );
        });
      }
      if (!props.hideInfo) {
        valueCells.alarmState = (
          <AlarmCell
            flags={flags}
            row={row}
            classes={props.classes}
            path={props.attribute.calculated.path}
            infoClickHandler={props.infoClickHandler}
            selectedRow={props.selectedRow}
          />
        );
      }
      return valueCells;
    };

    console.log('hello!');
    return (
      <div style={{ width: '100%', height: '100%' }}>
        <ReactDataGrid
          columns={columnHeadings}
          rowsCount={values.length}
          rowGetter={row => getRowData(row, columnWidgetTags, this.props)}
          minHeight={700}
        />
        {meta.writeable ? (
          <div>
            <IconButton
              onClick={() =>
                this.props.addRow(
                  this.props.attribute.calculated.path,
                  values.length
                )
              }
            >
              <Add style={{ width: '32px', height: '32px' }} />
            </IconButton>
          </div>
        ) : null}
        <div style={{ display: 'flex', position: 'bottom', padding: '4px' }}>
          {this.props.footerItems}
        </div>
      </div>
    );
  }
}
export default withStyles(styles, { withTheme: true })(NewWidgetTable);
