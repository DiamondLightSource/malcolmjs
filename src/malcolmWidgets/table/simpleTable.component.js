import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { emphasize } from '@material-ui/core/styles/colorManipulator';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

const styles = theme => ({
  rowFormat: {
    height: '30px',
  },
  textHeadings: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.3),
    textAlign: 'Center',
  },
  textBody: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.1),
    textAlign: 'Center',
  },
});

const SimpleTable = props => {
  const columnHeadings = props.labels.map(label => (
    <TableCell className={props.classes.textHeadings} padding="none">
      {label}
    </TableCell>
  ));
  const table = props.labels.map(label => props.values[label]);
  return (
    <Table>
      <TableHead>
        <TableRow className={props.classes.rowFormat}>
          {columnHeadings}
        </TableRow>
      </TableHead>
      <TableBody>
        {table[0].map((value, row) => (
          <TableRow className={props.classes.rowFormat}>
            {props.labels.map((label, column) => (
              <TableCell className={props.classes.textBody} padding="none">
                {table[column][row]}
              </TableCell>
            ))}
          </TableRow>
        ))}
      </TableBody>
    </Table>
  );
};

SimpleTable.propTypes = {
  labels: PropTypes.arrayOf(PropTypes.string).isRequired,
  values: PropTypes.shape({}).isRequired,
  classes: PropTypes.shape({
    textBody: PropTypes.string,
    textHeadings: PropTypes.string,
    rowFormat: PropTypes.string,
  }).isRequired,
};

export default withStyles(styles, { withTheme: true })(SimpleTable);
