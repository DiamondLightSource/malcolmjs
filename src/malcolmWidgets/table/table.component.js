/* eslint react/no-array-index-key: 0 */

import * as React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import { emphasize } from '@material-ui/core/styles/colorManipulator';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableFooter from '@material-ui/core/TableFooter';

import TableWidgetSelector, { getTableWidgetTags } from './widgetSelector';

const styles = theme => ({
  headerLayout: {
    tableLayout: 'fixed',
    width: 'calc(100% - 15px)',
  },
  headerLayoutNoScroll: {
    tableLayout: 'fixed',
    width: '100%',
  },
  tableLayout: {
    tableLayout: 'fixed',
  },
  tableBody: {
    overflowY: 'auto',
    height: 'calc(100% - 75px)',
    width: '100%',
  },
  rowFormat: {
    height: '30px',
  },
  textHeadings: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.3),
    textAlign: 'Center',
  },
  textBody: {
    backgroundColor: emphasize(theme.palette.background.paper, 0.1),
    textAlign: 'Center',
    padding: '2px',
  },
});

const WidgetTable = props => {
  const values =
    props.localState === undefined
      ? props.attribute.value
      : props.localState.value;
  const columnLabels = Object.keys(props.attribute.meta.elements);
  const rowChangeHandler = (rowPath, newValue) => {
    const rowValue = {};
    columnLabels.forEach(label => {
      rowValue[label] = values[label][rowPath.row];
      return 0;
    });
    rowValue[rowPath.label] = newValue;
    props.eventHandler(props.attribute.path, rowValue, rowPath.row);
  };

  const columnWidgetTags = getTableWidgetTags(props.attribute);
  const rowNames = values[columnLabels[0]];
  const columnHeadings = columnLabels.map((label, column) => (
    <TableCell
      className={props.classes.textHeadings}
      padding="none"
      key={column}
    >
      {label}
    </TableCell>
  ));
  return (
    <div>
      <Table
        className={
          rowNames.length > 20
            ? props.classes.headerLayout
            : props.classes.headerLayoutNoScroll
        }
      >
        <TableHead>
          <TableRow className={props.classes.rowFormat}>
            {columnHeadings}
          </TableRow>
        </TableHead>
      </Table>
      <div className={props.classes.tableBody}>
        <Table className={props.classes.tableLayout}>
          <TableBody>
            {rowNames.map((name, row) => (
              <TableRow className={props.classes.rowFormat} key={row}>
                {columnLabels.map((label, column) => (
                  <TableCell
                    className={props.classes.textBody}
                    padding="none"
                    key={[row, column]}
                  >
                    <TableWidgetSelector
                      columnWidgetTag={columnWidgetTags[column]}
                      value={values[label][row]}
                      rowPath={{ label, row, column }}
                      rowChangeHandler={rowChangeHandler}
                      columnMeta={props.attribute.meta.elements[label]}
                      setFlag={props.setFlag}
                    />
                  </TableCell>
                ))}
              </TableRow>
            ))}
          </TableBody>
        </Table>
      </div>
      <Table className={props.classes.headerLayout}>
        <TableFooter>
          <TableRow className={props.classes.rowFormat}>
            {props.footerItems.map((item, key) => (
              <TableCell key={key}>{item}</TableCell>
            ))}
          </TableRow>
        </TableFooter>
      </Table>
    </div>
  );
};

WidgetTable.propTypes = {
  attribute: PropTypes.shape({
    path: PropTypes.arrayOf(PropTypes.string),
    value: PropTypes.shape({}),
    name: PropTypes.string,
    typeid: PropTypes.string,
    labels: PropTypes.arrayOf(PropTypes.string),
    pending: PropTypes.bool,
    errorState: PropTypes.bool,
    alarm: PropTypes.shape({
      severity: PropTypes.number,
    }),
    meta: PropTypes.shape({
      label: PropTypes.string,
      tags: PropTypes.arrayOf(PropTypes.string),
      elements: PropTypes.shape({}),
    }),
    unableToProcess: PropTypes.bool,
  }).isRequired,
  localState: PropTypes.shape({
    value: PropTypes.shape({}),
  }),
  classes: PropTypes.shape({
    tableBody: PropTypes.string,
    headerLayoutNoScroll: PropTypes.string,
    headerLayout: PropTypes.string,
    tableLayout: PropTypes.string,
    textHeadings: PropTypes.string,
    textBody: PropTypes.string,
    rowFormat: PropTypes.string,
  }).isRequired,
  eventHandler: PropTypes.func.isRequired,
  // eslint-disable-next-line react/no-unused-prop-types
  setFlag: PropTypes.func.isRequired,
  footerItems: PropTypes.arrayOf(PropTypes.node),
};

WidgetTable.defaultProps = {
  localState: undefined,
  footerItems: [],
};
export default withStyles(styles, { withTheme: true })(WidgetTable);
