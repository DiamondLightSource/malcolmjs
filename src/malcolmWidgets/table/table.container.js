import * as React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { withTheme } from '@material-ui/core/styles';
import Typography from '@material-ui/core/Typography';
import ButtonAction from '../buttonAction/buttonAction.component';

import blockUtils from '../../malcolm/blockUtils';
import {
  malcolmUpdateTable,
  malcolmSetTableFlag,
  malcolmSetFlag,
  malcolmCopyValue,
  malcolmPutAction,
} from '../../malcolm/malcolmActionCreators';
import WidgetTable from './table.component';

const TableContainer = props => {
  const path = [props.blockName, props.attributeName];
  if (props.attribute.localState === undefined) {
    props.copyTable(path);
  }

  const footerItems = [
    ...props.footerItems,
    props.attribute.localState &&
    props.attribute.localState.flags.table.fresh ? (
      <Typography>Up to date!</Typography>
    ) : (
      <Typography>
        Update received @{' '}
        {`${new Date(props.attribute.raw.timeStamp.secondsPastEpoch * 1000)}`}
      </Typography>
    ),
    <ButtonAction
      clickAction={() => props.copyTable(path)}
      text="Discard changes"
    />,
    <ButtonAction
      clickAction={() => props.putTable(path, props.attribute.localState)}
      text="Submit"
    />,
  ];
  return (
    <WidgetTable
      attribute={props.attribute}
      localState={props.attribute.localState}
      eventHandler={props.eventHandler}
      setFlag={props.setFlag}
      footerItems={footerItems}
      addRow={props.addRow}
    />
  );
};

const mapStateToProps = (state, ownProps) => {
  let attribute;
  if (ownProps.attributeName && ownProps.blockName) {
    attribute = blockUtils.findAttribute(
      state.malcolm.blocks,
      ownProps.blockName,
      ownProps.attributeName
    );
  }
  return {
    attribute,
  };
};

const mapDispatchToProps = dispatch => ({
  eventHandler: (path, value, row) => {
    dispatch(malcolmUpdateTable(path, value, row));
  },
  addRow: (path, row) => {
    dispatch(malcolmUpdateTable(path, { insertRow: true }, row));
  },
  setFlag: (path, row, flagType, rowFlags) => {
    dispatch(malcolmSetTableFlag(path, row, flagType, rowFlags));
  },
  copyTable: path => {
    dispatch(malcolmCopyValue(path));
  },
  putTable: (path, tableState) => {
    const value = {};
    tableState.labels.forEach(label => {
      value[label] = [];
    });
    tableState.value.forEach(row => {
      tableState.labels.forEach(label => {
        value[label] = [...value[label], row[label]];
      });
    });

    dispatch(malcolmSetFlag(path, 'pending', true));
    dispatch(malcolmPutAction(path, value));
  },
});

TableContainer.propTypes = {
  blockName: PropTypes.string.isRequired,
  attributeName: PropTypes.string.isRequired,
  attribute: PropTypes.shape({
    localState: PropTypes.shape({
      value: PropTypes.shape({}),
      isDirty: PropTypes.bool,
      flags: PropTypes.shape({
        table: PropTypes.shape({
          dirty: PropTypes.bool,
          fresh: PropTypes.bool,
        }),
      }),
    }),
    raw: PropTypes.shape({
      timeStamp: PropTypes.shape({
        secondsPastEpoch: PropTypes.string,
      }),
    }),
  }).isRequired,
  eventHandler: PropTypes.func.isRequired,
  setFlag: PropTypes.func.isRequired,
  addRow: PropTypes.func.isRequired,
  copyTable: PropTypes.func.isRequired,
  putTable: PropTypes.func.isRequired,
  footerItems: PropTypes.arrayOf(PropTypes.node),
  /* theme: PropTypes.shape({
    palette: PropTypes.shape({
      primary: PropTypes.shape({
        light: PropTypes.string,
      }),
      background: PropTypes.shape({
        paper: PropTypes.string,
      }),
    }),
  }).isRequired, */
};

TableContainer.defaultProps = {
  footerItems: [],
};

export default connect(mapStateToProps, mapDispatchToProps)(
  withTheme()(TableContainer)
);
