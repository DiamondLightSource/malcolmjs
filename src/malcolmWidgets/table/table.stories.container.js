import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import SimpleTable from './simpleTable.component';
import WidgetTable from './table.component';

const styles = theme => ({
  testDiv: {
    backgroundColor: theme.palette.background.paper,
    width: 700,
    padding: 10,
    margin: 'none',
  },
});

const ContainedTable = props =>
  props.simple ? (
    <div className={props.classes.testDiv}>
      <SimpleTable
        labels={props.attribute.labels}
        values={props.attribute.values}
      />
    </div>
  ) : (
    <div className={props.classes.testDiv}>
      <WidgetTable
        attribute={props.attribute}
        eventHandler={props.eventHandler}
      />
    </div>
  );

ContainedTable.propTypes = {
  simple: PropTypes.bool.isRequired,
  attribute: PropTypes.shape({
    labels: PropTypes.arrayOf(PropTypes.string),
    values: PropTypes.shape({}),
  }).isRequired,
  classes: PropTypes.shape({
    testDiv: PropTypes.string,
  }).isRequired,
  eventHandler: PropTypes.func.isRequired,
};

export default withStyles(styles, { withTheme: true })(ContainedTable);
