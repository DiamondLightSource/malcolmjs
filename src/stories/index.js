import React from 'react';
import { configure, addDecorator } from '@storybook/react';
import { MuiThemeProvider, createMuiTheme } from '@material-ui/core/styles';
import { blue } from '@material-ui/core/colors';

const req = require.context('../', true, /\.stories\.js$/);

const theme = createMuiTheme({
  palette: {
    type: 'dark',
    primary: blue,
  },
  alarmState: {
    warning: '#e6c01c',
  },
});

const MaterialUIThemeDecorator = storyFn => (
  <MuiThemeProvider theme={theme}>{storyFn()}</MuiThemeProvider>
);
addDecorator(MaterialUIThemeDecorator);

function loadStories() {
  req.keys().forEach(filename => req(filename));
}

configure(loadStories, module);
