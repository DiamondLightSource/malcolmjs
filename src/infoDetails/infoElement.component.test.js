import React from 'react';
import configureStore from 'redux-mock-store';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import InfoElement from './infoElement.component';
import { harderAttribute } from '../malcolmWidgets/table/table.stories';

const state = {
  malcolm: {
    blocks: {
      test1: {
        attributes: [harderAttribute],
      },
    },
  },
};
describe('InfoElement', () => {
  let shallow;
  let mount;
  let mockStore;
  beforeEach(() => {
    harderAttribute.calculated.testStuff = {
      somethingFalse: false,
      somethingTrue: true,
      aString: 'testing!',
      aNumber: 1,
    };
    mockStore = configureStore();
    shallow = createShallow({ dive: true });
    mount = createMount();
  });

  it('renders correctly with value given', () => {
    const wrapper = shallow(
      <InfoElement
        key={1}
        label="Test element"
        value="testing..."
        alarm={null}
        tag="info:multiline"
        store={mockStore(state)}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('renders correctly with paths given', () => {
    const wrapper = shallow(
      <InfoElement
        key={1}
        blockName="test1"
        attributeName="layout"
        label="Test element"
        valuePath="calculated.testStuff.aString"
        alarmPath="calculated.testStuff.aNumber"
        tag="info:multiline"
        store={mockStore(state)}
      />
    );
    expect(wrapper.dive()).toMatchSnapshot();
  });

  it('button hooks up correctly', () => {
    const clickHandler = jest.fn();
    const wrapper = mount(
      <InfoElement
        key={1}
        label="Test element"
        value="test button"
        alarm={null}
        tag="info:button"
        handlers={{ clickHandler }}
        store={mockStore(state)}
      />
    );
    wrapper
      .find('button')
      .last()
      .simulate('click');
    expect(clickHandler).toHaveBeenCalledTimes(1);
  });

  const runDisabledButtonTest = extraProps => {
    const clickHandler = jest.fn();
    const wrapper = mount(
      <InfoElement
        key={1}
        blockName="test1"
        attributeName="layout"
        label="Test element"
        value="test button"
        alarm={null}
        tag="info:button"
        handlers={{ clickHandler }}
        store={mockStore(state)}
        {...extraProps}
      />
    );
    wrapper
      .find('button')
      .last()
      .simulate('click');
    expect(clickHandler).toHaveBeenCalledTimes(0);
  };

  it('button disables correctly', () => {
    runDisabledButtonTest({
      disabled: true,
    });
  });

  it('button disables correctly with flag path', () => {
    runDisabledButtonTest({
      disabledFlagPath: 'calculated.testStuff.somethingTrue',
    });
  });

  it('button disables correctly with flag path with NOT', () => {
    runDisabledButtonTest({
      disabledFlagPath: 'NOT.calculated.testStuff.somethingFalse',
    });
  });
});
