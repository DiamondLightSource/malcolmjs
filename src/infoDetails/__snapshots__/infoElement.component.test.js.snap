// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoElement renders correctly with paths given 1`] = `
<div
  className="InfoElement-div-1"
>
  <WithTheme(AttributeAlarm)
    alarmSeverity={1}
  />
  <WithStyles(Typography)
    className="InfoElement-textName-2"
  >
    Test element
  </WithStyles(Typography)>
  <div
    className="InfoElement-controlContainer-3"
    style={Object {}}
  >
    <WithStyles(WidgetTextUpdate)
      Text="testing!"
      noWrap={false}
    />
  </div>
</div>
`;

exports[`InfoElement renders correctly with value given 1`] = `
<div
  className="InfoElement-div-1"
>
  <WithTheme(AttributeAlarm)
    alarmSeverity={null}
  />
  <WithStyles(Typography)
    className="InfoElement-textName-2"
  >
    Test element
  </WithStyles(Typography)>
  <div
    className="InfoElement-controlContainer-3"
    style={Object {}}
  >
    <WithStyles(WidgetTextUpdate)
      Text="testing..."
      noWrap={false}
    />
  </div>
</div>
`;
